### PR TITLE
Add reduce function with two array input

### PR DESCRIPTION
**Pull request by rongrong**

### CODE REVIEW

**Comment by rongrong** [comment]
We could. This makes it slightly simpler to write I suppose.

**Comment by ArturGajowy** [comment]
I think we should refrain from broadening SQL user interface by adding this function then.

quoting @electrum:

> Adding anything that affects the SQL user interface such as new syntax, functions, etc., is the most scrutinized type of change since it is permanent (compared to internal things which can be easily refactored). We recommend filing issues for new, non-standard functions so the interface can be reviewed before writing the code.

>It doesn't make sense to add custom functions for things that are trivially composable using existing functions, unless there is a very compelling reason.

**Comment by haozhun** [comment]
I agree with @ArturGajowy that we need to be careful when introducing new functions. The reason we intend to add such a function is primarily driven by performance. If we use zip, an intermediate value of type `array<row<T1,T2>>` will have to be produced, which is quite expensive. We didn't use a quantitive benchmark to substantiate this claim. And we should. 

@rongrong, can you please add a benchmark to compare this function with zip+reduce. That benchmark will also be useful later when we improve the efficiency of these functions.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayReduceTwoInputsFunction.java [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+    public static final ArrayReduceTwoInputsFunction ARRAY_REDUCE_TWO_INPUTS_FUNCTION = new ArrayReduceTwoInputsFunction();
+
+    private static final MethodHandle METHOD_HANDLE = methodHandle(ArrayReduceTwoInputsFunction.class, "reduce", Type.class, Type.class, Block.class, Block.class, Object.class, MethodHandle.class, MethodHandle.class);
+
+    private ArrayReduceTwoInputsFunction()
+    {
+        super(new Signature(
+                "reduce",
+                FunctionKind.SCALAR,
+                ImmutableList.of(typeVariable("T"), typeVariable("Q"), typeVariable("S"), typeVariable("R")),
```
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayReduceTwoInputsFunction.java]: Compared to `T` and `Q`, `T` and `U` is a more common choice. You can also do `T1` and `T2` if that's what you prefer.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayReduceTwoInputsFunction.java [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+    public ScalarFunctionImplementation specialize(BoundVariables boundVariables, int arity, TypeManager typeManager, FunctionRegistry functionRegistry)
+    {
+        Type leftInputType = boundVariables.getTypeVariable("T");
+        Type rightInputType = boundVariables.getTypeVariable("Q");
+        Type intermediateType = boundVariables.getTypeVariable("S");
+        Type outputType = boundVariables.getTypeVariable("R");
+        MethodHandle methodHandle = METHOD_HANDLE.bindTo(leftInputType).bindTo(rightInputType);
+        return new ScalarFunctionImplementation(
+                true,
+                ImmutableList.of(false, true, true, false, false),
```
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayReduceTwoInputsFunction.java]: `U` should not be nullable

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayReduceTwoInputsFunction.java [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+            Object leftInput = readNativeValue(leftInputType, leftBlock, position);
+            Object rightInput = readNativeValue(rightInputType, rightBlock, position);
+            try {
+                intermediateValue = inputFunction.invoke(intermediateValue, leftInput, rightInput);
+            }
+            catch (Throwable throwable) {
+                throw Throwables.propagate(throwable);
+            }
+        }
+        // If the length of the two arrays are different, use null as the rest of elements in the shorter one.
```
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayReduceTwoInputsFunction.java]: For now, let's fail invocations where left and right don't match in length.

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/TestArrayReduceTwoInputsFunction.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    }
+
+    @Test
+    public void testMultipleValueState()
+        throws Exception
+    {
+        assertFunction(
+                "reduce(" +
+                        "ARRAY [5, 20, 50], ARRAY [3, 6, 9]," +
+                        "CAST(ROW(0, 0, 0) AS ROW(sumA BIGINT, sumB BIGINT, count INTEGER)), " +
```
Comment (haozhun) [inline presto-main/src/test/java/com/facebook/presto/operator/scalar/TestArrayReduceTwoInputsFunction.java]: It's not a bit deal. But weighted average feels like a more interesting test/example.

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/TestArrayReduceTwoInputsFunction.java [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+            throws Exception
+    {
+        assertFunction("reduce(ARRAY [], ARRAY [], CAST (0 AS BIGINT), (s, x, y) -> s + x + y, s -> s)", BIGINT, 0L);
+    }
+
+    @Test
+    public void testBasic()
+            throws Exception
+    {
+        assertFunction("reduce(ARRAY [5, 20, 50], ARRAY [1, 2, 3], CAST (0 AS BIGINT), (s, x, y) -> s + x * y, s -> s)", BIGINT, 195L);
```
Comment (haozhun) [inline presto-main/src/test/java/com/facebook/presto/operator/scalar/TestArrayReduceTwoInputsFunction.java]: Make sure your test covers cases where `left` and `right` have different types.

**Comment by ArturGajowy** [comment]
Closing as this seems abandoned given the time of last activity.

**Pull request closed without merging**